lib/spack/spack/build_systems/compiler.py

```diff
@@ -0,0 +1,126 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+import itertools
+import os
+import re
+from typing import Sequence
+
+import llnl.util.tty as tty
+from llnl.util.lang import classproperty
+
+import spack.compiler
+import spack.package_base
+
+
+class CompilerPackage(spack.package_base.PackageBase):
+    # metadata identifying this as a compiler for lmod
+    family = "compiler"
+
+    # Optional suffix regexes for searching for this type of compiler.
+    # Suffixes are used by some frameworks, e.g. macports uses an '-mp-X.Y'
+    # version suffix for gcc.
+    suffixes: Sequence[str] = [r"-.*"]
+
+    # Optional prefix regexes for searching for this compiler
+    prefixes: Sequence[str] = []
+
+    #: Compiler argument that produces version information
+    version_argument = "-dumpversion"
+
+    #: Return values to ignore when invoking the compiler to get its version
+    ignore_version_errors: Sequence[int] = ()
+
+    #: Regex used to extract version from compiler's output
+    version_regex = "(.*)"
+
+    #: Platform matcher for Platform objects supported by compiler
+    is_supported_on_platform = lambda x: True
+
+    #: Static definition of languages supported by this class
+    compiler_languages = ["c", "cxx", "fortran"]
+
+    #: Dynamic definition of languages supported by this package
+    @property
+    def supported_languages(self):
+        return self.compiler_languages
+
+    # TODO: how do these play nicely with other tags
+    tags = ["compiler"]
+
+    @classproperty
+    def compiler_names(cls):
+        names = []
+        for language in cls.compiler_languages:
+            names.extend(getattr(cls, f"{language}_names"))
+        return names
+
+    @classproperty
+    def executables(cls):
+        regexp_fmt = r"^({0}){1}({2})$"
+        return [
+            regexp_fmt.format(prefix, re.escape(name), suffix)
+            for prefix, name, suffix in itertools.product(
+                cls.prefixes + [""], cls.compiler_names, cls.suffixes + [""]
+            )
+        ]
+
+    @classmethod
+    def determine_version(cls, exe):
+        try:
+            output = spack.compiler.get_compiler_version_output(exe, cls.version_argument)
+            match = re.search(cls.version_regex, output)
+            if match:
+                return ".".join(match.groups())
+        except spack.util.executable.ProcessError:
+            pass
+        except Exception as e:
+            tty.debug(e)
```

Comment:
We should probably leave a better trace on the provenance of the error, and not just dump the error message

Reply:
Will commit a better message.